compare initial master

MatiPoli opened the pull request.

juampi74 committed:
Corregí un error en el html de marcas

juampi74 committed:
Acorté los nombres de los archivos ...-created-or-modified.service porque al clonar de cero el repositorio, por la longitud del nombre de esos archivos, no los traía bien

juampi74 committed:
Acorté más los nombres

juampi74 committed:
Acorté los nombres de las carpetas de cada entidad

MatiPoli committed:
Prueba para GitHub pages

MatiPoli committed:
Prueba para github pages 2

MatiPoli committed:
Github Pages cancelado

MatiPoli committed:
Primera versión del sistema de autentificación de usuarios

MatiPoli committed:
Segunda versión del sistema de autentificación de usuarios

Añadido cookies y passwords encriptadas

MatiPoli committed:
Navegación de angular actualizada

MatiPoli committed:
Login y Register hechos

MatiPoli committed:
Auth Guard hecho y funcionando

MatiPoli committed:
Configuración del proxy en el ng serve

MatiPoli committed:
Intento de logout

MatiPoli committed:
Intento de logout 2

MatiPoli committed:
Intento de logout 3

MatiPoli committed:
Revert "Intento de logout 3"

This reverts commit d305e56fc8975b192a472bb56915a7e0df976cba.

Marcos-Godoy committed:
abm de vehiculos hecho con tabla funcionadndo

Marcos-Godoy committed:
toqueteo el back para q me agregue la imagen del auto

MatiPoli committed:
Logout completo y interceptor http creado

MatiPoli committed:
Revert "toqueteo el back para q me agregue la imagen del auto"

This reverts commit 63d09fb39a70807538dd300211f0f29af66094c3.

MatiPoli committed:
Merge branch 'main' into Login

MatiPoli committed:
Merge pull request #4 from DSW-3k02-GGJM/Login

Integración de avances de Login

MatiPoli committed:
Interceptor funcionando

Marcos-Godoy committed:
agrego manejo de imagen, no anda del todo bien

Marcos-Godoy committed:
logrue que se cargue una ruta en la bd

Marcos-Godoy committed:
tratando de conectar con la card

Marcos-Godoy committed:
cuando doy de alta un auto, aparece la foto en las cards para alquilar

Marcos-Godoy committed:
hago q las cards sean por modelo y pongo el boton de alquilar sin funcionalidad

MatiPoli committed:
Primera refactorización al Inglés

Marcos-Godoy committed:
arreglo cosas

GilardoniLucio committed:
Backend traducido

GilardoniLucio committed:
Backend traducido

GilardoniLucio committed:
traduccion parcial frontend

GilardoniLucio committed:
terminando

GilardoniLucio committed:
terminando

GilardoniLucio committed:
terminando v2

GilardoniLucio committed:
terminando v3

GilardoniLucio committed:
traduccion comlpeta

GilardoniLucio committed:
Merge branch 'Ingles'

GilardoniLucio committed:
Solucioné el tema de los modelos

- Solucioné el tema de los modelos
- Borré unas carpetas del BE que habían quedado de más
- Renombré cosas por consistencia

GilardoniLucio committed:
Solucione el issue que estaba en github

- Solucioné el issue que estaba en Github
- Encontré mas carpetas para borrar

MatiPoli committed:
Login, Register y Logout más interactivo

MatiPoli committed:
Manejo de errores

Detecta cuando ya está en uso el mail (sólo frontend)

GilardoniLucio committed:
##Agregué orm.ts al gitignore##

Para no tener que cambiar el client url cada vez. Tener en cuenta si lo van a modificar

GilardoniLucio committed:
Añadí sanitizeInputs faltantes

na04ber committed:
CRUD Reservas por parte del admin (faltan algunas funciones y validaciones)

Marcos-Godoy committed:
empiezo a hacer el filtro

MatiPoli committed:
Merge branch 'main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App

MatiPoli committed:
Mejor manejo de errores

MatiPoli committed:
Sistema de roles en el backend hecho

MatiPoli committed:
FUNCIONAAAAAAAAAA

Marcos-Godoy committed:
implemento el query builder me da errores en vehicle controller verrr

Marcos-Godoy committed:
arreglo errores q me tiraba pero sigue sin andarrrrr

GilardoniLucio committed:
Lo de buscar por x anda nuevamente

El filtro ahora toma como tercer argumento el campo por el cual se busca
Ademas cambie algunas 'o's por '0's para que el ide no los marque como TODOs

GilardoniLucio committed:
Merge branch 'main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App

GilardoniLucio committed:
Revert "Merge branch 'main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App"

This reverts commit e209ff4c1b71d9929501e652b1254485afb3dde9, reversing
changes made to d04a9d76561d6b4b6042ceef0d770c028cc03206.

GilardoniLucio committed:
Reapply "Merge branch 'main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App"

This reverts commit d01aa44d248db38d0f499c7e0fdeb024df83a5ad.

GilardoniLucio committed:
quickFix

MatiPoli committed:
Mejoras menores en sistema login

MatiPoli committed:
Usuario admin se crea si no existe

MatiPoli committed:
Sistema de guardas para roles hecho

juampi74 committed:
Primera version Check-in y Check-out Reserva

Marcos-Godoy committed:
cambios

na04ber committed:
Arreglo que al crear Reserva solo aparezcan las opciones de tipoDoc y numero doc correctos

GilardoniLucio committed:
Funcionalidad Minima

solo anda lo de select *

GilardoniLucio committed:
Anda solo desde el back

Falta:
- conectar con front
- añadir seleccion sucursal
- que la query considere aquellos cuyas reservas hayan sido canceladas
- me debo estar olvidando de algo
- le metemos un filtro post query por categoria/cant pasajeros/tipo transmision?

GilardoniLucio committed:
Funciona

Falta:
- lo que dice en el commit anterior
- validaciones form
- form modal?
- el resto del proceso

MatiPoli committed:
Cambio de estética

Marcos-Godoy committed:
pantalla de alquiler funcionando pero como si fuera de admin

Marcos-Godoy committed:
no me abre el modal cuando aprieto el alquilar pero despues lo sigo

MatiPoli committed:
Arreglo texto en blanco

GilardoniLucio committed:
Backend definitivo

- añadí validaciones
- la response ahora debería ser igual al resto

GilardoniLucio committed:
Backend Definitivo v2

- borré cosas que sobraban
- tdv no me animo a borrar el codigo comentado en app.ts. Lo dejo para despues del merge

MatiPoli committed:
Input de angular material

GilardoniLucio committed:
Backend definitivo v3

Me estuve peleando con el calendario

Marcos-Godoy committed:
sigo metiendole todavia no logro q ande

Marcos-Godoy committed:
se abre el modal

MatiPoli committed:
Instalo Jest

MatiPoli committed:
Revert "Instalo Jest"

This reverts commit 356d9c3762de3a8ff49dc9e348fff0ad2aa7c113.

Marcos-Godoy committed:
creo un service para pasar la fecha todavia no funca

Marcos-Godoy committed:
cambio el service por un  output

MatiPoli committed:
Nuevo estilo + cambiando register form

MatiPoli committed:
Mas cambios en el register form

MatiPoli committed:
forms de login y register hechos

GilardoniLucio committed:
Merge branch 'filtroDeAlquiler' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App into filtroDeAlquiler

GilardoniLucio committed:
Mergeado asi nomas. No anda (naturalmente)

revisar
vehicle-list.component.ts
vehicle-filter.component.ts

MatiPoli committed:
Perfil de usuario hecho

MatiPoli committed:
Guard y permisos para ver perfiles hecho

MatiPoli committed:
Formulario de marca actualizado

MatiPoli committed:
Formulario de color actualizado

MatiPoli committed:
Formulario de categorias actualizado

MatiPoli committed:
Formulario de sucursal actualizado

MatiPoli committed:
Gestion de users casi completa

MatiPoli committed:
bugfix

MatiPoli committed:
bugfix2

Validador agregado

MatiPoli committed:
bugfix 3

Las fechas se muestran bien

MatiPoli committed:
Menu de admin / empleado hecho

MatiPoli committed:
bugfix4

MatiPoli committed:
Formulario de modelo actualizado

Falta hacer bien el tema de las imágenes

MatiPoli committed:
Formulario de vehiculo actualizado

MatiPoli committed:
Formulario de reserva actualizado

Capaz se me escapa algo

MatiPoli committed:
bugfix5

MatiPoli committed:
Vereificaciones de location, brand, category y color en backend

Capaz falta alguna

MatiPoli committed:
bugfix

MatiPoli committed:
Verificaciones de user, vehicle, vehicle model y reservation hechas

MatiPoli committed:
Hecho los entityName-exists de cada entidad para verificar que no exista

MatiPoli committed:
Bugfix

MatiPoli committed:
Hecho mostrar error al eliminar entidades usadas

MatiPoli committed:
Hechas las validaciones para que no se repitan nombre

MatiPoli committed:
Bugfic

MatiPoli committed:
Mejorada la barra de búsqueda de cada entidad

GilardoniLucio committed:
quickfix

ahora anda

GilardoniLucio committed:
el filtro no anda

en el backend anda bien. Actualmente la request sale vacia (no se si está bien).
Hay logging de consola con comportamiento esperado.
Hay mucho código comentado para borrar.

GilardoniLucio committed:
Merge branch 'filtroDeAlquiler'

GilardoniLucio committed:
ahora compila

arreglados errores que impedían que compile
Crashea cuando quiero ver vehiculos desde staff

GilardoniLucio committed:
Crash solucionado

Marcos-Godoy committed:
hago una reserva con valores hardcodeados cunado aprieto alquilar

Marcos-Godoy committed:
basicamente deje comentada unas cosas para probar despues porque me tengo q ir

MatiPoli committed:
bugfix

GilardoniLucio committed:
Soluciones varias

Anda filtro alquiler*
Conversión de client a user
Otros

GilardoniLucio committed:
Ver Descripcion

- Mejorada lógica filtros
- Eliminado requerimiento de autenticación de las sucursales
- Las cards aparecen con contenido
- Agregada opcion distinct (inactivo)
- TODO: Revisar estilos cards

MatiPoli committed:
Bugfix

Marcos-Godoy committed:
anda la reserva hardcodeada

Marcos-Godoy committed:
paso fechas

Marcos-Godoy committed:
aaaaaaaaaaaaa

MatiPoli committed:
Bugfixes y nuevo crear reserva

MatiPoli committed:
Merge branch 'main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App

MatiPoli committed:
Mejorado el responsive

na04ber committed:
Cambio en el filtro de disponibilidad de un vehiculo para un alquiler

juampi74 committed:
Arreglado Check-In y Check-Out, y realizado el cálculo del precio final de una reserva según el precio por día de la categoría del modelo del vehículo asociado y la cantidad de días de la reserva

na04ber committed:
Camnio fitro para que se pueda filtrar por atributos anidados y agrego filtro por documentID en las reservas admin

juampi74 committed:
Deshabilito botones de editar y eliminar reservas para las que ya fue realizado el Check-Out. Corrijo datos accedidos en la lista de vehiculos filtrados para que devuelva bien los vehículos disponibles

MatiPoli committed:
Varios Cambios

Semi finalizado el curso normal del cliente. Cambié los modales de boostrap por los de angular material porque son más lindos. La reserva ahora se crea con un vehículo cualquiera pero en teoría está todo preparado para la query de lucio.

También eliminé initialkms y totalkms de las reservas ya que no funca.

MatiPoli committed:
Bugfix

na04ber committed:
Vuelvo a agregar initialKms y finalKms porque se usan para el checkIn y el checkOut.  Elimino boton de editar en las reservas admin.

na04ber committed:
Elimino las opciones de editar en el formulario de reservas admin.

na04ber committed:
Arreglo error con el form reservas admin

juampi74 committed:
Agrego cálculo de precio final para una nueva reserva

Marcos-Godoy committed:
cambios en los estilos de la tarjetitaaaaa

juampi74 committed:
Modificada la vista de admin para crear una nueva reserva, tomando como base el stepper de la vista cliente

juampi74 committed:
Arreglo errores check-in y check-out causados por el commit anterior

juampi74 committed:
Corrijo error en el reservation.routes.ts

juampi74 committed:
Modifico el filtro de vehiculos disponibles para una nueva reserva, para que devuelva un vehiculo por cada modelo (el primero que haya)

juampi74 committed:
Actualizo imagen del modelo de datos en el README

MatiPoli committed:
Bugfix

MatiPoli committed:
Merge branch 'main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App

juampi74 committed:
Actualizo README con instrucciones de instalación y ejecución del proyecto

juampi74 committed:
Le otorgo permisos al empleado para que pueda acceder a la lista de usuarios del sistema al momento de la creación de una nueva reserva

juampi74 committed:
Merge branch 'main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App

MatiPoli committed:
Sistema para enviar emails hecho

Hace falta poner el Email y la Password correspondiente en el archivo .env

Para esto comunicarse conmigo asi les paso el código.

MatiPoli committed:
Mejor manejo de errores del mailer

juampi74 committed:
Agrego link al video de demostracion de la aplicacion en el README

MatiPoli committed:
Update README.md

juampi74 committed:
Actualizo instrucciones de instalacion y ejecucion del proyecto

juampi74 committed:
Actualizo formato del README

juampi74 committed:
Agrego informacion al README

GilardoniLucio committed:
Limpieza rápida de código

Borrados spec, recli folder, cli folder

Co-Authored-By: Matias Tomás Marquez <70172645+MatiPoli@users.noreply.github.com>

GilardoniLucio committed:
Solucion parcial validación fecha nacimiento***

Sigue sin andar

Co-Authored-By: Matias Tomás Marquez <70172645+MatiPoli@users.noreply.github.com>

GilardoniLucio committed:
Merge branch 'main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App

GilardoniLucio committed:
Revert "Solucion parcial validación fecha nacimiento***"

This reverts commit dd9e2bf424f0bc8db18957424897319c15d05011.

MatiPoli committed:
Bugfix y agregado de campo verificado (no hace nada)

MatiPoli committed:
Update README.md

GilardoniLucio committed:
Añadidas varias validaciones al front

GilardoniLucio committed:
Merge branch 'validaciones'

GilardoniLucio committed:
Reintento validaciones front

Ahora con éxito*
si el dni/otro no se ingresa bien de entrada, hay que apretar "guardar" dos veces

GilardoniLucio committed:
Merge branch 'reintento-validaciones'

juampi74 committed:
Actualizo README con una aclaracion sobre la importacion de datos de prueba a traves de un script

MatiPoli committed:
Sistema de verificación de usuario hecho

Ahora los usuarios necesitan verificar su email para pode logearse.

MatiPoli committed:
Bugfix

MatiPoli committed:
Sistema de cambio de contraseña hecho

MatiPoli committed:
Cambios los modales de error y confirmación de eliminar por dialog

Los dialog de check in y check out habría que cambiarlos por unos más acordes

MatiPoli committed:
Termino de eliminar los modales viejos

MatiPoli committed:
Bugfix de issue

MatiPoli committed:
Sistema para comunicarse con los usuarios

Faltan corregir detalles de la vista mobile

MatiPoli committed:
bugfix

GilardoniLucio committed:
Update README.md

Había un error en el archivo sql con datos de prueba no permitía iniciar sesión al admin

juampi74 committed:
Agrego filtro por fecha para las reservas en la vista de admin

juampi74 committed:
Agrego vista para que el cliente pueda visualizar sus reservas

GilardoniLucio committed:
Arreglos mínimos

Marcos-Godoy committed:
implemente una api para subir la imagen, me falta el front

Marcos-Godoy committed:
Merge branch 'temporal-main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App into temporal-main

Marcos-Godoy committed:
borro algo q no iba

GilardoniLucio committed:
Modificación pop-up confirmación de eliminación

Entidades afectadas: Marcas, Categorias, Colores, Sucursales, Usuarios, Vehiculos, Modelos
Antes no mostraba icono, ni 'Eliminar' en el botón
Además se cambiaron algunos colores para que sea consistente con la eliminación de reservas

aotaduy commented on backend/src/app.ts:
```diff
@@ -0,0 +1,62 @@
+import dotenv from 'dotenv';
+
+dotenv.config();
+const frontendURL = (process.env.FRONTEND_DOMAIN || 'http://localhost') + (process.env.FRONTEND_PORT || ':4200');
```
no esta mal pero genralmente no es lo mejor servir el front desde el mismo server node que el back

aotaduy commented on backend/src/brand/brand.controler.ts:
```diff
@@ -0,0 +1,169 @@
+      res.status(404).json({ message: 'The brand does not exist' });
+    } 
+    else {
+      res.status(200).json({ message: 'The brand has been found', data: brand });
```
tal vez en los casos de exito no es necesario el mensaje se puede devolver la informacion dir3ectaente aunque no esta mal tampoco

aotaduy commented on backend/src/brand/brand.controler.ts:
```diff
@@ -0,0 +1,169 @@
+  try {
+    const brand = em.create(Brand, req.body.sanitizedInput);
+    await em.flush();
+    res.status(201).json({ message: 'The brand has been created', data: brand });
```
201 quiere decir no content no deberia devolver nada

aotaduy commented on backend/src/brand/brand.routes.ts:
```diff
@@ -0,0 +1,15 @@
+brandRouter.get('/:id', AuthService.isAuthenticated(["admin"]), findOne)
+brandRouter.post('/', AuthService.isAuthenticated(["admin"]), sanitizedBrandInput , add)
+brandRouter.put('/:id', AuthService.isAuthenticated(["admin"]), sanitizedBrandInput , update)
+//brandRouter.patch('/:id', AuthService.isAuthenticated(["admin"]), sanitizedBrandInput , update)
```
no dejemos codigo comentado

aotaduy commented on backend/src/category/category.controler.ts:
```diff
@@ -0,0 +1,178 @@
+          'vehicleModels.brand',
+          'vehicleModels.vehicles',
+          'vehicleModels.vehicles.color',
+          'vehicleModels.vehicles.location',
```
este mismo array lo veo repetido en vaarios lugares lo podriamos centralizar en un archivo relacionado con el modelo

aotaduy commented on backend/src/reservation/reservation.controler.ts:
```diff
@@ -0,0 +1,299 @@
+  const user = req.body.sanitizedInput.user;
+  const vehicle = req.body.sanitizedInput.vehicle;
+
+  console.log('Datos de la reserva:', req.body.sanitizedInput);
```
eliominemos console log superfluos

aotaduy commented on backend/src/reservation/reservation.controler.ts:
```diff
@@ -0,0 +1,299 @@
+
+export {
+  sanitizedReservationInput,
+  sanitizedAdminReservationInput,
```
necesitas exportar todos? los inpout tambien ?

aotaduy commented on backend/src/category/category.routes.ts:
```diff
@@ -0,0 +1,16 @@
+categoryRouter.get('/', AuthService.isAuthenticated(["admin"]), findAll);
+categoryRouter.get('/:id', AuthService.isAuthenticated(["admin"]), findOne);
+categoryRouter.post('/', AuthService.isAuthenticated(["admin"]), sanitizedCategoryInput, add);
+categoryRouter.put('/:id', AuthService.isAuthenticated(["admin"]), sanitizedCategoryInput, update);
```
podriamos buscar uan forma de hacer sanitized mas generico ? asi no lo tenemos que agrtegar manualmenter en cada ruta ?

aotaduy commented on backend/src/middleware/entityManager.middleware.ts:
eliminar archivos vacios

aotaduy commented on backend/src/user/user.controller.ts:
```diff
@@ -0,0 +1,584 @@
+
+dotenv.config();
+
+const SECRET_KEY = process.env.SECRET_KEY || 'Aca va una clave secretisima que está publicada en github usea que tan secreta no era';
```
esta ya esta en el archivo auth.service deberiamos encapsular el acceso a los secret en un solo lugar

aotaduy commented on backend/src/user/user.controller.ts:
```diff
@@ -0,0 +1,584 @@
+    /*
+    //const token = AuthService.generateToken(usuario); // Crea un token y lo asocia al usuario
+    const { password: _, ...publicUser} = user;
+    const token = AuthService.generateToken(user, SECRET_KEY, '1h'); // Crea un token y lo asocia al usuario
```
sino funciona eliminemos veo que ene l login esta

aotaduy commented on backend/src/user/user.controller.ts:
```diff
@@ -0,0 +1,584 @@
+        else {
+          const isValid = await bcrypt.compare(password, user.password)
+          if(!isValid){
+            return res.status(401).json({ message: 'Wrong password' });
```
no conviene decir por separado si el password e correcto o si el email es correcto para dificultar ataques por fuerza bruta es mejor unificar nun solo mensaje de login incorrecto sino primero adivinan el email ydespues el password y es mas facil. 
o es mas facil verificar si alguien es usuario del sitio probado con un password incorrecto

aotaduy commented on frontend/src/app/app.component.scss:
```diff
@@ -0,0 +1,38 @@
+
+footer {
+  background-color: #0F222D ;
+  color: #FFF5D0;
```
usemos variables para los colores

aotaduy commented on frontend/src/app/app.component.ts:
```diff
@@ -0,0 +1,28 @@
+  ngOnInit() {
+    this.authService.isAuthenticated().subscribe(
+      res=> {
+        console.log("res");
```
eliminemos console log innecesarios

aotaduy commented on frontend/src/app/brand-folder/brand-form/brand-form.component.ts:
```diff
@@ -0,0 +1,118 @@
+  }
+
+  onSubmit() {
+    if(!this.brandForm.invalid) {
```
tal vez habria que deshbilitar el boton en el form en vez de esto ya que puede desorientar al usuario

aotaduy commented on frontend/src/app/brand-folder/brand-created-or-modified/brand.service.ts:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  notifyBrandCreatedOrModified() {
+    this.brandCreatedOrModified.next();
```
para que es esta notificacion ?

juampi74 replied:
Implementamos notificaciones de ese estilo porque necesitábamos una forma de que el componente de la tabla, en este caso de Marcas, se entere de que se insertó una nueva marca o se modificó una existente, y recargue la información de la tabla.
No lo podíamos manejar con la emisión de eventos como está hecho para las eliminaciones, porque el formulario en el que se crean y editan las entidades (Marcas) no es un componente hijo del componente llamado brands, que es el encargado de obtener todas las marcas y pasárselas a la tabla para que las muestre.
En cambio, en el caso de la eliminación, esta se hace directamente desde la tabla de marcas, que es un componente hijo del componente brands mencionado, por lo que lo pudimos resolver mediante la emisión de un evento.
En resumen, utilizamos ese servicio para pasar información entre distintos componentes que no están relacionados de forma directa (formulario de creacion/edicion y componente brands).

aotaduy commented on frontend/src/app/brand-folder/brands/brands.component.ts:
```diff
@@ -0,0 +1,68 @@
+
+  onBrandDeleted(brandId: number): void {
+    // Maneja la eliminación de una marca
+    this.brands = this.brands.filter((brand) => brand.id !== brandId);
```
si bien puede hacerce como optimizacion seria correcto tambien recargar las marcas del server ya que puede haber otros usuarios editando al mismo tiempo. Se puede hacer de esta forma como un cache optimista pero cada tanto hay que recargar del server

aotaduy commented on frontend/src/app/form-floating-textfield/form-floating-textfield.component.ts:
```diff
@@ -0,0 +1,15 @@
+import { Component, Input } from '@angular/core';
+
+@Component({
+  selector: 'app-form-floating-textfield',
```
se usa este componente ? no lo vi en los forms

aotaduy commented on frontend/src/app/form-floating-select/form-floating-select.component.ts:
```diff
@@ -0,0 +1,14 @@
+import { Component, Input } from '@angular/core';
+
+@Component({
+  selector: 'app-form-floating-select',
```
se usa este ? donde? sino borremoslo

aotaduy commented on frontend/src/app/guards/auth-employee.guard.ts:
```diff
@@ -0,0 +1,30 @@
+  return authService.isAuthenticated().pipe(
+    switchMap(() => {
+      return authService.checkEmployee().pipe(
+        map(() => {
```
hay mucho codigo duplicado con los guard no se podria simplificar pasando la callback que cambia ?

aotaduy commented on frontend/src/app/home-page/home-page.component.scss:
```diff
@@ -0,0 +1,148 @@
+}
+
+.hero {
+  color: #ffffff;
```
usemos variables para los colores o en este caso podria ser white como color nombrado pero mejor variables

aotaduy commented on frontend/src/app/home-page/home-page.component.scss:
```diff
@@ -0,0 +1,148 @@
+}
+
+.hero h1 {
+  font-size: 2.5em;
```
hay ems y rems porque ?

aotaduy commented on frontend/src/app/interceptors/auth.interceptor.ts:
```diff
@@ -0,0 +1,20 @@
+export const authInterceptor: HttpInterceptorFn = (req, next) => {
+  const authService = inject(AuthService) as AuthService;
+  return next(req).pipe(
+    tap(() => {
```
sino hace nada borrenoslo en este caso no hace nada

juampi74 committed:
Agrego opcion de cancelar reserva para el cliente, arreglo filtro de vehiculos disponibles para una nueva reserva y centralizo el acceso al string de conexion a la base de datos

MatiPoli committed:
Cambios en el generic success dialog

MatiPoli committed:
Merge branch 'temporal-main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App into temporal-main

MatiPoli committed:
Corregido error ortográfico

MatiPoli committed:
Bugfix

MatiPoli committed:
Bugfix 2

aotaduy commented on frontend/src/app/service/api.service.ts:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  getOne(entity: string, id: number): Observable<any> {
+    return this.http.get(`${this.apiUrl}/${entity}/${id}`, {
```
mismo para todos los endpoints donde es dinamico el tipo que se pide y su respuesta

aotaduy commented on frontend/src/app/service/api.service.ts:
```diff
@@ -0,0 +1,110 @@
+
+  create(entity: string, data: any): Observable<any> {
+    const headers = new HttpHeaders({
+      'Content-Type': 'application/json',
```
con httpclient no debeerian necesitar especifcar el content type por defecto es json

aotaduy commented on frontend/src/app/vehicleModel-folder/vehicleModel-created-or-modified/vehicleModel.service.ts:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  set IsDataLoaded(value: boolean) {
+    this.isDataLoaded = value;
```
veo mucha duplicacion en servicios similares a este, no podriamos crear uno solo ?

aotaduy commented on frontend/src/app/vehicleModel-folder/vehicleModel-created-or-modified/vehicleModel.service.spec.ts:
```diff
@@ -0,0 +1,16 @@
+    TestBed.configureTestingModule({});
+    service = TestBed.inject(VehicleModelCreatedOrModifiedService);
+  });
+
```
sino implementan realmente tests mejor eliminenlos

aotaduy commented on frontend/src/app/vehicle-folder/vehicle-list/vehicle-list.component.ts:
```diff
@@ -0,0 +1,90 @@
+  fetchVehicles(filter: any) {
+    this.http
+      .get<any>(
+        `/api/vehicles/available?startDate=${filter.startDate}&endDate=${filter.endDate}&location=${filter.location}`
```
esto deberia estar encapsulado en servicios, no directo dfesde aca

aotaduy commented on frontend/src/app/vehicle-folder/vehicle-list/vehicle-list.component.ts:
```diff
@@ -0,0 +1,90 @@
+        (response) => {
+          console.log('Response data:', response);
+          this.response = response.data.map((vehicle: any) => {
+            console.log('Vehicle data:', vehicle);
```
eliminemos todos los console.log que no son necesarios

aotaduy commented on frontend/src/app/user-folder/users/users.component.ts:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  onUserDeleted(userId: number): void {
+    this.users = this.users.filter(
```
lo mismo de anges sobre recargar la pagina

aotaduy commented on frontend/src/app/user-folder/validation-dialog/validation-dialog.component.ts:
```diff
@@ -0,0 +1,23 @@
+  styleUrl: './validation-dialog.component.scss'
+})
+export class ValidationDialogComponent {
+
```
para que tenemos una clase vacia?

aotaduy commented on frontend/src/app/user-folder/user-form/user-form.component.ts:
```diff
@@ -0,0 +1,190 @@
+      } else {
+        this.action = 'Create';
+        this.title = 'Nuevo usuario';
+        this.buttonText = 'Registrar';
```
podriamos armar un objeto en una variable con todas estas propiedaddes

aotaduy commented on frontend/src/app/user-folder/users-table/users-table.component.html:
```diff
@@ -0,0 +1,60 @@
+        <td>{{ user.verified }}</td>
+        <td>
+          <div class="d-flex justify-content-around">
+            <button class="btn btn-dark fs-5" (click)="editUser(user)">
```
este esta en todas las trablas tal vez podriamos armar un componente

aotaduy commented on frontend/src/app/vehicleModel-folder/vehicleModels-table/vehicleModels-table.component.ts:
```diff
@@ -0,0 +1,71 @@
+import { MatInputModule } from '@angular/material/input';
+
+@Component({
+  selector: 'app-vehicleModels-table',
```
mantengamos la convencion de nombres en este caso esta mezclado camelCase con dash-case

MatiPoli committed:
Bugfix 3

MatiPoli commented on frontend/src/app/cat-folder/categories-table/categories-table.component.ts:
```diff
@@ -0,0 +1,71 @@
+                if (error.status === 400) { 
+                  const modalRef = this.modalService.open(GenericErrorModalComponent);
+                  modalRef.componentInstance.title = 'Error al eliminar la categoría';
+                  modalRef.componentInstance.message = 'La categoría no se puede eliminar porque tiene modelos asociados.';
```
Los errores están muy controlados en el back. Por lo que no se rompería. Pero si, lo mejor sería que se mande directamente desde el backend.

GilardoniLucio committed:
[Closes #11] Redirije a listado de reservas si la reserva la hace staff

GilardoniLucio committed:
Unifiqué estilos de tablas

Además restringí el ancho de la columna de editar/eliminar

na04ber committed:
Agrego boton de cancelar reserva en la vista del admin y agrego que se deshabilite las opciones de cancelar/eliminar una vez se hizo el check in

MatiPoli committed:
Sistema de mandar recordatorios automáticos hecho

MatiPoli committed:
Merge branch 'temporal-main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App into temporal-main

na04ber committed:
deshabilito botones checkIn y cancelar dependiendo las fechas de inicio o fin de la reserva

MatiPoli committed:
Bugfix

MatiPoli committed:
Merge branch 'temporal-main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App into temporal-main

Marcos-Godoy committed:
anda el backend

Marcos-Godoy committed:
Merge branch 'temporal-main' of https://github.com/DSW-3k02-GGJM/TP-3k02-Berli-Gilardoni-Godoy-Jaca-Marquez-App into temporal-main

MatiPoli committed:
Bugfix

Marcos-Godoy committed:
funcionando, faltan detalles

Marcos-Godoy committed:
Funcionando...

Marcos-Godoy committed:
hicimos para poder eliminar las fotos viejas

Marcos-Godoy committed:
hago limpieza y paso algunas cosas a ingels

juampi74 committed:
Reorganización de la estructura de carpetas, limpieza y refactorización del código

juampi74 committed:
Merge de temporal-main a main

juampi74 committed:
Elimino populates innecesarios en los controladores y el metodo findOne de Reservations (que no se utilizaba)

juampi74 committed:
Deshabilito botones Submit en los formularios cuando los datos ingresados no son válidos

juampi74 committed:
Mejoro las validaciones relacionadas con fechas (fecha de nacimiento de usuarios y rango de fechas para reservas)

juampi74 committed:
Unifico el uso de ems y rems en los estilos (dejando todo en rems)

juampi74 committed:
Armo un api.service específico para cada entidad, eliminando el genérico. Defino interfaces para mejorar la seguridad de tipos en los retornos HTTP, dejando de usar 'any'

Además:
- Elimino el 'Content-Type' en los headers de las solicitudes HTTP, ya que con HttpClient el valor por defecto ya es JSON
- Generalizo todos los guards en uno solo, en el que se manejan todos los roles
- Elimino el auth.interceptor que no se utilizaba
- Generalizo los generic-dialogs en uno solo que sirva tanto para los casos de éxito como de error (elimino también el componente ValidationDialogComponent, ya que ahora utilizo este nuevo generic-dialog)

juampi74 committed:
Obtengo mensajes de error al eliminar una entidad directamente desde el backend

Además:
- Arreglo error al obtener las entidades relacionadas en los formularios de vehículo y modelo (al momento de la edición)
- Arreglo errores al cargar una nueva imagen para un modelo, al editar una existente, y al eliminarla (momento en el que se elimina el modelo)

En el commit anterior, sumado a lo ya dicho:
- Recargo las entidades desde la base de datos  al eliminar alguna, en lugar de eliminar esta última de la lista de entidades en memoria

juampi74 committed:
Armo un nuevo componente reutilizable que contiene los botones para editar y eliminar una entidad. Actualizo los archivos README.md, y el archivo '.mysql' con datos de prueba

Además:
- Actualizo la imagen del modelo de datos en el README.md principal
- Agrego archivo de ejemplo para el '.env' del backend ('.env.example')
- Organizo la carpeta '/assets' del frontend
- Elimino algunas imágenes que ya no se utilizan de la carpeta '/assets' tanto del backend como de la raíz del proyecto

juampi74 committed:
Implemento variables para los colores en las hojas de estilo

Además:
- Hago algunos retoques para mejorar el responsive tanto en la página principal como en las tarjetas de los modelos de vehículos al momento de realizar una reserva

juampi74 committed:
Centralizo la información de los formularios en una sola variable tipada con las propiedades correspondientes

juampi74 committed:
Generalizo los sanitizedInput del backend en una única función que se encarga de realizar las validaciones correspondientes según los parámetros que recibe

Además:
 - Evito que los mensajes de error en los formularios se solapen sobre otros elementos en pantallas chicas
 - Cierro barras de navegación al navegar en pantallas chicas
 - Modifico el estilo de los botones para agregar una nueva entidad, para que sean acordes al estilo del resto de la página
 - Agrego algunas validaciones más en campos de los formularios, principalmente en campos relacionados a los nombres de las entidades
 - Elimino algunas hojas de estilo .scss que generaban código duplicado, ya que el mismo ya estaba definido en las hojas genéricas de estilo
 - Unifico hojas de estilo de los steppers en los que tanto los admins/empleados como los clientes realizan las reservas, dejando una sola que aplique para ambos componentes

juampi74 committed:
Agrego funcionalidad al momento del check-out, para que sea necesario indicar el kilometraje actual del vehículo y si se tiene que devolver o no el depósito al cliente

Con ese objetivo:
 - Agrego un campo finalPrice a la tabla de reservas, en el que se registra el precio final de la misma, teniendo en cuenta tanto la cantidad de días como el precio por día y el depósito en garantía

Además:
 - Modifico la estructura del archivo .env, dejándolo como se puede ver en el archivo .env.example, para facilitar la configuración de las direcciones URL en las que se ejecutan tanto el backend como el frontend
 - Modifico el archivo config.ts del backend para que las validaciones que realiza sean acordes a lo realizado en el ítem anterior
 - Hago algunos retoques tanto en las tablas como en los formularios, para mejorar el responsive
 - Adapto algunos mensajes que se le muestran al usuario, para que sean acordes al estilo en el que están redactados otros mensajes
 - Unifico "servicios" de cálculo del precio de las reservas, para que en uno solo se encuentren las funciones que calculan tanto el precio inicial al momento de realizar la reserva como el precio final al momento del check-out de la misma (Para el cálculo final se tiene en cuenta lo descripto en el primer ítem)

juampi74 committed:
Unifico todos los campos de fechas de la base de datos a tipo DATE (en reservas y recordatorios) y obtengo todos los mensajes mostrados al usuario directamente desde el backend

Con esos objetivos:
 - Realizo las modificaciones necesarias para evitar problemas relacionados a zonas horarias en el almacenamiento de las fechas
 - Traduzco los mensajes devueltos por la API del backend a español, para mostrarle dichos mensajes directamente al usuario en el frontend

Además:

BACKEND

 - Elimino imágenes de la carpeta '/public' que no se correspondían con el formato en el que actualmente se almacenan las imágenes en dicha carpeta
 - Agrego una validación adicional a la hora de realizar una nueva reserva, para que antes de registrarla vuelva a confirmar que el vehículo sigue disponible (en caso contrario se ejecuta un rollback y se le muestra el mensaje correspondiente de error al usuario en el frontend)
- Separo la lógica del Check-Out de una reserva de la de un simple update, para poder manejar correctamente la actualización, tanto de la reserva como del vehículo, dentro de una transacción
 - Modifico la lógica de eliminación de las reservas, para que solo sea posible en el caso de que todavía no se le haya mandado el recordatorio al cliente
 - Recorto espacios exteriores en el valor recibido en los métodos de verificación de unicidad antes de comparar, para mejorar la validación
 - Convierto el documentID de los usuarios a mayúsculas antes de guardarlo, para unificar el formato de los pasaportes que contienen letras
 - Solo devuelvo mensaje de 'Error al eliminar la imagen' de un modelo si no se trata de que la misma no fue encontrada, para mejorar la experiencia del usuario al no mostrarle un error innecesario

FRONTEND

 - Centralizo la apertura del modal genérico GenericDialogComponent a un servicio compartido (OpenDialogService), cuyos métodos utilizan el resto de los componentes
 - Unifico la forma de mostrar los mensajes de error para que los mismos se muestren a través del modal genérico (salvo en casos específicos), redireccionando a distintas rutas al momento del cierre del modal según corresponda
 - Modularizo el código TypeScript de los componentes en varias funciones separadas, para que, por ejemplo, no se lleve a cabo todo en el ngOnInit( ), onSubmit( ), entre otras
 - Defino funciones específicas 'handleSubmit( )', 'handleError( )' y algunas otras de nombres similares, para que manejen las distintas respuestas provenientes del backend
 - Le agrego el modificador de acceso 'private' a aquellos métodos que sólo son accedidos desde el propio código TypeScript del componente en el que están definidos
 - Paralelizo solicitudes para obtener información del backend en ReservationFormComponent y ReservationStepperComponent, para mejorar el rendimiento
 - Elimino código redundante en el archivo HTML de EmailVerificationComponent, manejado ahora con los mensajes dinámicos devueltos por el backend
 - Renombro los métodos 'clickEvent( )' utilizados para modificar la visibilidad de los campos relacionados a contraseñas, a 'togglePasswordVisibility( )'
 - Elimino dicho método y las variables e importaciones correspondientes en el ProfileComponent, dado que, al no tener su formulario un campo de contraseña, no se utilizaban
 - Elimino archivo 'date-formats.config.ts' ya no utilizado
 - Modifico el tamaño de algunas de las imágenes guardadas en '/assets' dentro del frontend, para que sean más acordes al tamaño en el que son mostradas, y así evitar posibles errores

GENERAL

 - Modifico la imagen del modelo de datos, para que sea acorde a los cambios mencionados sobre los campos de fechas y al agregado del campo final_price en la tabla de reservas realizado en el commit anterior
 - Modifico algunos otros pequeños detalles en el README.md principal

juampi74 committed:
Modifico el tipo de los campos de precios de la tabla 'category' de la base de datos a INT (para que sea acorde a como están manejados en el sistema)

En conjunto con eso:
 - Agrego validaciones en los formularios que contienen campos de 'type=number' para corroborar que lo ingresado sea un número entero (formularios de Categorías, Vehículos, Modelos, y en el modal genérico cuando se realiza el Check-Out de una Reserva)
 - Modifico el middleware sanitizedInput del backend para que permita ingresar 0 en los campos numéricos (de modo de poder registrar un vehículo 0 KM)
 - Modifico la imagen del modelo de datos y el archivo .mysql con datos de prueba, para que se correspondan con los cambios realizados

Además:
 - Vuelvo a programar los recordatorios de reservas cada 10 minutos (en lugar de cada 1 minuto), ya que fue modificado por error en el commit anterior

juampi74 committed:
Unifico el manejo de fechas para que en todo el sistema se utilice la librería externa date-fns

Una vez hecho eso:
 - Elimino las librerías de manejo de fechas luxon y moment de las dependencias del backend, ya que no se utilizan más

Además:
 - Elimino la librería mysql no utilizada del backend
 - Modifico algunos detalles extra en los package.json del backend y del frontend
 - Acomodo pequeños detalles para mejorar el responsive en los steppers de reservas

na04ber committed:
Agrego spec.ts del vehicle-form con algunos tests

juampi74 committed:
Refactorizo el archivo de tests unitarios vehicle-form.component.spec.ts

Con el objetivo de mejorar la legibilidad, organización y mantenibilidad de los tests, y garantizar el correcto funcionamiento del formulario en diferentes situaciones:
 - Elimino imports innecesarios
 - Defino función reutilizable setFormValues() para establecer los valores del formulario
 - Extraigo código duplicado al método beforeEach(), que se ejecuta antes de cada uno de los tests
 - Organizo los tests, utilizando varios describe() para agrupar aquellos relacionados entre sí
 - Agrego más tests para comprobar el funcionamiento del formulario en diferentes escenarios, incluyendo casos límite y tipos de datos incorrectos

na04ber committed:
Agrego una primera versión de un test E2E con cypress en el frontend

juampi74 committed:
Modifico el archivo cypress.config.ts para solucionar un error al ejecutar el Test E2E

GilardoniLucio committed:
Documentación

1er borrador de documentación api

na04ber committed:
Test E2E terminado

MatiPoli committed:
Documentación de user

juampi74 committed:
Tests para el Backend terminados

Además:
 - Actualizo formato de la documentación en general
 - Agrego sección de autenticación y errores en la documentación de la API

juampi74 committed:
Actualizo el Video de Demostración de Funcionamiento de la Aplicación

Además:
 - Elimino archivos .http que se utilizaban en su momento para testear manualmente los endpoints, ya que los mismos no cuentan con la autenticación necesaria
 - Los tests de backend realizados cubren ya la mayoría de los endpoints disponibles en nuestra API

juampi74 committed:
Agrego instrucciones de ejecución de los tests automáticos

juampi74 committed:
Actualizo imagen y formato en las instrucciones de ejecución de los tests automáticos

juampi74 committed:
Última actualización de formato

juampi74 committed:
Ordeno tabulaciones en dichas instrucciones

juampi74 committed:
Actualizo formato de varios archivos Markdown de documentación

juampi74 committed:
Arreglo links en las instrucciones de instalación

juampi74 committed:
Agrego Evidencia de Ejecución de Tests Automáticos

juampi74 committed:
Agrego Minutas de Reunión y Avance

Además:
 - Organizo la estructura de carpetas de la documentación

juampi74 committed:
Actualizo formato de Minutas de Reunión y Avance

GilardoniLucio committed:
Redacté docs de varias entidades

Detalle: category, color, location, reservation, vehicle

juampi74 committed:
Documentación de la API del backend terminada

Detalles:
 - Agrego dcoumentación para la entidad Modelo
 - Actualizo documentación de la entidad Usuario, para unificar el formato con la del resto de las entidades

juampi74 committed:
Actualizo formato en parte de la Documentación de la API del backend

juampi74 committed:
Hago la última actualización de formato en la Documentación de la API del backend

juampi74 committed:
Corrijo parte de la Documentación de la API del backend

juampi74 committed:
Elimino ítem actualmente no utilizado en el índice de la documentación del proyecto

juampi74 committed:
Corrijo error de ortografía en las instrucciones de ejecución de tests automáticos

aotaduy commented on backend/__tests__/api.test.ts:
```diff
@@ -0,0 +1,2197 @@
+        'El vehículo no se puede eliminar porque tiene reservas asociadas.',
+    });
+  });
+
```
el archivo de tests conviene separarlo y ponerlo junto con las apis que esta testeando, no un archivo unico

juampi74 replied:
Entendemos que los tests deberían poder funcionar de forma independiente, pero decidimos implementarlos de esta forma para poder probar secuencialmente los endpoints de las entidades relacionadas.

Por ejemplo, probamos el endpoint de creación de una categoría, después el de una marca, y con esa categoría y marca que creamos, probamos el endpoint de creación de un modelo, que es dependiente de las dos primeras entidades.

aotaduy commented on backend/src/config.ts:
```diff
@@ -0,0 +1,53 @@
+  NODE_ENV: z
+    .enum(['test', 'development', 'production'])
+    .default('development'),
+  BACKEND_URL: z.string().url().default('http://localhost:3000/'),
```
los url deberian ser configurables para cuando lo deployen en otro lado podrian usar el paquete 'env'

aotaduy commented on backend/src/config.ts:
```diff
@@ -0,0 +1,53 @@
+  BACKEND_URL: z.string().url().default('http://localhost:3000/'),
+  FRONTEND_URL: z.string().url().default('http://localhost:4200/'),
+  SECRET_KEY: z.string().default('password1'),
+  SECRET_EMAIL_KEY: z.string().default('password2'),
```
todos estos deberian ser configurables por variables de entorno

juampi74 replied:
Cada uno de esos valores son configurables a través de variables de entorno en un archivo .env que siga la forma del archivo de ejemplo .env.example que se encuentran en la raíz del proyecto backend.

En este archivo config.ts, se le dan valores por defecto en caso de que las variables de entorno no sean definidas.

Además, se realizan validaciones utilizando la librería Zod, para comprobar que los valores dispuestos para las variables de entorno sean acordes a lo esperado.

aotaduy commented on backend/src/core/brand/brand.controller.ts:
```diff
@@ -0,0 +1,108 @@
+  try {
+    const id = Number.parseInt(req.params.id);
+    const brand = await em.findOne(Brand, { id });
+    const brandInUse = await em.findOne(VehicleModel, { brand: id });
```
estas restricciones estan en la base de datos tambien?

aotaduy replied:
seria bueno usar integridad referncial a nivel de la bd mas alla de este control

juampi74 replied:
Sí, en la base de datos, si se trata de eliminar un registro en una tabla que está referenciado a través de claves foráneas por un registro de otra tabla, existe una restricción de integridad que no lo permite.

Un ejemplo de esto es si se trata de eliminar un registro que representa a una marca que está referenciado por un registro que representa a un modelo. En ese caso, se lanza una restricción de integridad por parte de la base de datos.

Además, con respecto a la unicidad del nombre, también hay una restricción de Unique en la columna brandName, que garantiza que no se puedan registrar dos marcas con el mismo nombre. La existencia de un índice en esta columna se puede ver reflejado en la tabla de nuestra aplicación en la que se muestra la información de las marcas, al ver que las mismas se ordenan alfabéticamente.

aotaduy commented on backend/src/core/brand/brand.controller.ts:
```diff
@@ -0,0 +1,108 @@
+    const id = Number.parseInt(req.params.id);
+    const brand = await em.findOneOrFail(Brand, { brandName });
+    if (brand.id === id) {
+      res.status(200).json({ exists: false });
```
se podrias simplificar la logica 
 res.status(200).json({ exists: brand.id === id});
(opcional)

juampi74 replied:
Es cierto, se ahorrarían algunas líneas de código con esa simplificación de la lógica.

juampi74 replied:
Solucionado, ya simplificamos la lógica en todos los casos de verificación de unicidad, ya sea de nombre o de patente.

aotaduy commented on backend/src/core/category/category.entity.ts:
```diff
@@ -0,0 +1,32 @@
+  depositValue!: number;
+
+  @OneToMany(() => VehicleModel, (vehicleModel) => vehicleModel.category, {
+    cascade: [Cascade.ALL],
```
mismo de antes, tal vez el cascade no es lo mejor

juampi74 replied:
Lo cambiaremos para indicar restricción al eliminar registros referenciados por otros como mencioné arriba.

juampi74 replied:
Solucionado. Como mencioné arriba, eliminamos todo lo relacionado a la eliminación en cascada en la definición de las entidades/tablas de nuestro esquema con MikroORM.

aotaduy commented on backend/src/shared/utils/format-date.ts:
```diff
@@ -0,0 +1,11 @@
+export const fromDashToSlash = (date: string): string => {
```
para las fechas les recomiendo date-fns en npm (opcional)

juampi74 replied:
Utilizamos tanto en el backend como en el frontend la librería date-fns para el manejo de fechas en general.

Se puede visualizar en el package.json de ambos partes del proyecto que está incluída dicha librería en las dependencias, y que se utiliza en múltiples casos a lo largo del desarrollo del proyecto.

Hubo casos en los que no encontramos la solución para lo que queríamos hacer utilizando la librería, como en el caso específico marcado acá de transformar una fecha del formato yyyy-mm-dd al formato dd/mm/yyyy, por lo que desarrollamos una función propia específica para resolver ese requerimiento.

Recuerdo que al querer utilizar una función de la librería para resolver dicho objetivo, teníamos problemas con las zonas horarias, ya que, a pesar de no contar con hora en nuestra fecha, la conversión restaba 3 horas a la fecha con 00:00:00 para transformarla de UTC a la zona horaria Argentina, quedando la fecha como de un día anterior a las 21:00:00, lo que nos daba incorrectamente la fecha que necesitábamos.

aotaduy commented on docs/pages/api/entities/brand/endpoints/create.md:
```diff
@@ -0,0 +1,117 @@
+
+## Descripción
+
+Este endpoint permite crear una marca.
```
estos docs los generaron con algo? son muy completos

juampi74 replied:
No, seguimos en primera instancia la forma en la que está documentada la API de Mercado Pago, pero a partir de eso lo documentamos nosotros adaptándolo a lo que necesitábamos para cubrir toda la información de los endpoints de nuestra API.

aotaduy commented on frontend/src/app/app.module.ts:
```diff
@@ -0,0 +1,45 @@
+    provideRouter(routes),
+    provideAnimationsAsync(),
+    AuthService,
+    { provide: MAT_DATE_LOCALE, useValue: 'en-GB' },
```
esta bien este locale ? no se puede usar el de aca?

juampi74 replied:
Tuvimos varios problemas con el manejo de fechas a lo largo del desarrollo de nuestra aplicación, y cuando encontramos la forma en la que nos funcionó correctamente lo dejamos así.

Posiblemente se pueda configurar distinto y encontrar la forma de que funcione, pero decidimos dejarlo funcionando una vez que pudimos solucionar los problemas que teníamos principalmente con las zonas horarias.

aotaduy commented on frontend/src/app/core/brand/components/brands-table/brands-table.component.ts:
```diff
@@ -0,0 +1,110 @@
+
+  get filteredBrands(): Brand[] {
+    return this.brands.filter((brand: Brand) =>
+      brand.brandName.toLowerCase().includes(this.filterRows.toLowerCase())
```
tienen algu filtro implementado en el backend?

juampi74 replied:
El único filtro que implementamos desde el lado del backend es aquel con el que filtramos los vehículos al momento de realizar una reserva, para hallar aquellos disponibles según un rango de fechas y una sucursal elegidos por el usuario. 

El mismo se puede ver en el controlador de la entidad vehicle en el backend. 

Para los filtros de búsqueda de entidades en las tablas que muestran la información, decidimos implementar un filtrado en memoria desde el lado del frontend una vez que ya recuperamos la totalidad de las entidades registradas en la base de datos.

aotaduy commented on frontend/src/app/core/category/components/categories-table/categories-table.component.html:
```diff
@@ -0,0 +1,62 @@
+    <input
+      matInput
+      type="text"
+      [(ngModel)]="filterRows"
```
evitemos el uso de ngModel lo recomendado es reactive forms

juampi74 replied:
Solo utilizamos ngModel en el caso del input de búsqueda que posibilita el filtrado en memoria que menciono en el comentario de arriba.

Para los formularios de creación y modificación de entidades, utilizamos en todos los casos Reactive Forms.

juampi74 replied:
Solucionado, reemplazamos el uso de ngModel por Reactive Forms en los campos de búsqueda que permiten realizar un filtrado sobre las entidades existentes según un campo específico.

aotaduy commented on frontend/src/app/core/category/components/categories-table/categories-table.component.ts:
```diff
@@ -0,0 +1,112 @@
+  ) {}
+
+  private openDeleteDialog(name: string, id: number): void {
+    const dialogRef: MatDialogRef<GenericDialogComponent, boolean> =
```
veo un poco repetido el tema del dialog, no e podria abstraer un poco y agregando unas callback o parametros hacer algo reutilizable? 
(opcional)

juampi74 replied:
Justamente implementamos reutilización de código abstrayendo la apertura de los modales genéricos (mat-dialogs) en un solo archivo './frontend/src/app/shared/services/notifications/open-dialog.service.ts'.

Lo hicimos lo más genérico posible contemplando los principales modales que solemos utilizar, para reducir al máximo los parámetros que le tenemos que pasar a dichas funciones que se encargan de la apertura.

Los parámetros que le pasamos desde los componentes que requieren abrir modales, son los mínimos necesarios que son específicos para el caso en el que se va a utilizar el modal en dicho componente.

juampi74 commented on backend/src/core/brand/brand.controller.ts:
```diff
@@ -0,0 +1,108 @@
+      res.status(200).json({ exists: true });
+    }
+  } catch (error) {
+    res.status(200).json({ exists: false });
```
Decidimos implementarlo de esta forma para no mostrar un mensaje de error al usuario en los formularios si la comprobación de unicidad fallaba por error de conexión con la base de datos.

En cambio, si se da ese error de conexión, la validación pasa, pero al apretar el botón submit de 'Registrar' se vuelve a validar la unicidad del campo en cuestión mediante una función genérica que sanitiza la entrada del usuario.

En ese caso, si el valor era duplicado, se le muestra un mensaje de error al usuario de que 'Se ha detectado un valor duplicado que debería ser unico'. Si vuelve a haber un error de conexión al momento del submit, ahí sí se le muestra un modal al usuario de que hubo un error de conexión, y la operación no se realiza. 

Como es una validación intermedia asincrónica que desencadena en un posible mensaje de error al usuario, nos pareció que esta era la mejor forma de implementarlo para mejorar la experiencia del usuario.

juampi74 commented on backend/src/core/brand/brand.entity.ts:
```diff
@@ -0,0 +1,23 @@
+  brandName!: string;
+
+  @OneToMany(() => VehicleModel, (vehicleModel) => vehicleModel.brand, {
+    cascade: [Cascade.ALL],
```
Si, significa eliminación en cascada. Por lo que veo nos quedó implementado de esa forma desde un principio, y después no volvimos a revisar la definición del esquema.

A pesar de eso, en nuestro sistema lo tenemos manejado con restricción de integridad que impide eliminar registros que están referenciados por otros mediante claves foráneas.

Lo corregiremos igualmente para que quede acorde a como está manejado en el sistema.

juampi74 replied:
Solucionado, eliminamos todo lo referente a la eliminación en cascada en la definición de todas las entidades/tablas del esquema con MikroORM, para que sea acorde a como está manejada la eliminación en el sistema (con restricción de integridad que impide la eliminación).

juampi74 committed:
Simplifico la lógica de verificación de unicidad de nombre/patente en los controladores en el backend

juampi74 committed:
Elimino lo referente a la eliminación en cascada en la definición de todas las entidades/tablas del esquema con MikroORM, para que sea acorde a como está manejada la eliminación en el sistema (con restricción de integridad que impide la eliminación)

juampi74 committed:
Reemplazo el uso de ngModel por Reactive Forms en los campos de búsqueda que permiten realizar un filtrado sobre las entidades existentes según un campo específico
